tests/lib/prepare.sh

```diff
@@ -14,29 +14,34 @@ update_core_snap_with_snap_exec_snapctl() {
     snap="$(mount | grep " $core" | awk '{print $1}')"
     umount --verbose "$core"
 
-    # Now unpack the core, inject the new snap-exec and snapctl into it, and
-    # repack it.
+    # Now unpack the core, inject the new snap-exec/snapctl into it
     unsquashfs "$snap"
     cp /usr/lib/snapd/snap-exec squashfs-root/usr/lib/snapd/
     cp /usr/bin/snapctl squashfs-root/usr/bin/
+    # also add snap/snapd because we re-exec by default.
```

Comment:
Indeed. From the current re-exec snap-confine is missing. Mostly because we need to think about confinement for that (the current approach won't work). However if we do pick which snap-confine to use in snapd itself we can easily get this fix in via a new core snap (as the snapd from the core snap will select which snap-confine to run).

As a side-note. Running snap-exec/snapctl from classic is right now hard because we are already inside the core snap at this point. But we could write code that somehow bind mounts them across (which I think has its own problems because how do we know if we must bind mount more than just the binary).

I think ideally we would have a single "snapd" snap that contains all this code and we don't re-exec at all but use the snap everywhere. On classic the deb of snapd would just contain a /usr/bin/snap "loader" that downlaods the snapd (or core) snap and then runs it. but obviously quite a bit more work and we need to think about transitions etc.